src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/CodeActionEndpoint.cs

```diff
@@ -0,0 +1,135 @@
+
+        public CodeActionEndpoint(
+            IEnumerable<RazorCodeActionProvider> providers,
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver,
+            ILoggerFactory loggerFactory)
+        {
+            if (providers is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (foregroundDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _providers = providers;
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+            _logger = loggerFactory.CreateLogger<CodeActionEndpoint>();
+        }
+
+        public CodeActionRegistrationOptions GetRegistrationOptions()
+        {
+            return new CodeActionRegistrationOptions()
+            {
+                DocumentSelector = RazorDefaults.Selector
+            };
+        }
+
+        public async Task<CommandOrCodeActionContainer> Handle(CodeActionParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+
+            if (document is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var sourceText = await document.GetTextAsync();
+            var linePosition = new LinePosition((int)request.Range.Start.Line, (int)request.Range.Start.Character);
+            var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
+            var location = new SourceLocation(hostDocumentIndex, (int)request.Range.Start.Line, (int)request.Range.Start.Character);
+
+            var context = new RazorCodeActionContext(request, request.TextDocument.Uri, codeDocument, location);
+            var tasks = new List<Task<CommandOrCodeActionContainer>>();
+
+            // Check CodeDocument.Kind == Kind.Component or .Legacy  .IsComponent()
+            if (!FileKinds.IsComponent(context.Document.GetFileKind()))
```

Comment:
This is a concern of each code action. For instance, it'd be totally acceptable for one of us to write a code action that applied to non-Blazor files.